Add video embed support for updates

SudharakaP opened the pull request:
Related to https://github.com/opencollective/opencollective/issues/4298

Depends on https://github.com/opencollective/opencollective-frontend/pull/6394

Betree commented on server/lib/notifications.js:
```diff
@@ -215,9 +215,87 @@ const notifyUpdateSubscribers = async activity => {
 };
 
+function replaceVideosByImagePreviews(activity) {
+  const regexImg = new RegExp(`<iframe([\\w\\W]+?)[\\/]?>`, 'g');
+  const regexUrl = new RegExp(
+    `(https):\\/\\/([\\w_-]+(?:(?:\\.[\\w_-]+)+))([\\w.,@?^=%&:/~+#-]*[\\w@?^=%&/~+#-])?`,
+    'ig',
+  );
+  activity.data.update.html = activity.data.update.html
+    .replace(regexImg, match => {
+      let imageReplacement;
+      if (match.includes('youtube')) {
+        imageReplacement = match
+          .replace(new RegExp('<iframe', 'g'), '<img')
+          .replace(new RegExp('allowfullscreen', 'g'), '')
+          .replace(new RegExp('frameborder="0"', 'g'), '');
+
+        const { videoService, videoId } = getEmbedDetails(imageReplacement);
+        const previewImageURL = constructPreviewImageURL(videoService, videoId);
+        return imageReplacement.replace(regexUrl, previewImageURL);
+      }
+    })
+    .replace(new RegExp('<figure data-trix-content-type="([\\w\\W]+?)[\\/]?">', 'g'), '')
+    .replace(new RegExp('</iframe><figcaption></figcaption></figure>', 'g'), '')
+    .replace(new RegExp('width="100%" height="394"', 'g'), '');
```
It's quite complex to undestand this code without proper context. It also looks like `.replace(new RegExp('<figure data-trix-content-type="([\\w\\W]+?)[\\/]?">', 'g'), '')` can have side effects that are not related to iframes.

I haven't tested the following code (other than the regex with https://regex101.com/r/OiYxtK/1), but I think using a slightly more advanced regex for the iframe can really simplify things here:

```es6
const iframePreviewRegex = /<iframe\s[^>]*src="([^"]+)"[^>]*><\/iframe>/ig;
activity.data.update.html = activity.data.update.html.replaceAll(iframePreviewRegex, (_, href) => {
  const { service, id } = parseServiceLink(href);
  const imgSrc = constructPreviewImageURL(service, id);
  if (imgSrc) {
    return `<img src="${imgSrc}" alt="${service} content">`
  } else {
    return ''
  }
})
```

SudharakaP replied:
@Betree : Thanks much. I've refactored this based on your suggestion. Seems to work perfectly. Now as for the `.replace(new RegExp('<figure data-trix-content-type="([\\w\\W]+?)[\\/]?">', 'g'), '')`, yes you are right; this is an oversight on my part and it could have side effects. 🙄 I removed this line as it's really not needed to remove these `<figure` tags anyways. 🐨

Betree reviewed (commented):
Would be great to have a preview for anchor.fm, people will be surprised if it just disappears from the content when they send emails.

Also, we really need unit tests to make sure future regressions on this code don't break the emails.

SudharakaP commented on server/lib/notifications.js:
```diff
@@ -215,9 +215,87 @@ const notifyUpdateSubscribers = async activity => {
+    .replace(new RegExp('<figure data-trix-content-type="([\\w\\W]+?)[\\/]?">', 'g'), '')
+    .replace(new RegExp('</iframe><figcaption></figcaption></figure>', 'g'), '')
+    .replace(new RegExp('width="100%" height="394"', 'g'), '');
+  console.log(activity.data.update.html);
```
Ops I've missed this 🤯

SudharakaP commented:
> Would be great to have a preview for anchor.fm, people will be surprised if it just disappears from the content when they send emails.
> 
> Also, we really need unit tests to make sure future regressions on this code don't break the emails.

@Betree : Working on this. I don't think we could get a preview image for anchor.fm as in YouTube. So I am gonna put generic image there. Hope that's adequate for this purpose. 🤔 🐨

SudharakaP commented:
@Betree : I've added the anchor preview as well. It looks looks like this. Let me know your thoughts. I wonder if this is the best placeholder to show but I felt this is good for now? 🤔 🐨 

![image](https://user-images.githubusercontent.com/12435965/119563696-be2a0d80-bd5c-11eb-9be2-df3856fa6409.png)

Betree commented:
> > Would be great to have a preview for anchor.fm, people will be surprised if it just disappears from the content when they send emails.
> > Also, we really need unit tests to make sure future regressions on this code don't break the emails.
> 
> @Betree : Working on this. I don't think we could get a preview image for anchor.fm as in YouTube. So I am gonna put generic image there. Hope that's adequate for this purpose. thinking koala

Yes, absolutely! Ideally, it should be something that lets you know you can click on it to read the track but for a first version any image just to make sure the content doesn't disappear is fine :wink:

Betree commented on server/lib/notifications.js:
```diff
@@ -215,9 +215,62 @@ const notifyUpdateSubscribers = async activity => {
+      } else {
+        return '';
+      }
+    })
```
![image](https://user-images.githubusercontent.com/1556356/119616640-203c4e80-be01-11eb-9116-d24b7a7b62ee.png)


Regarding this warning, I would add a small line to remove all iframes that we haven't been able to convert

```suggestion
    }).replace(/<iframe\s[^>]><\/iframe>/gi, '')
```

Betree replied:
We can also use `sanitize-html` for this

SudharakaP replied:
Done. 👍🏽 Didn't notice this warning in CI. 🤔

Betree commented on server/lib/notifications.js:
```diff
@@ -215,9 +215,62 @@ const notifyUpdateSubscribers = async activity => {
+        return '';
+      }
+    })
+    .replace(new RegExp('</iframe><figcaption></figcaption></figure>', 'g'), '')
+    .replace(new RegExp('width="100%" height="394"', 'g'), '');
```
I don't think we still need these two?

SudharakaP replied:
Nope, you are right. That's my mistake. 😞

SudharakaP commented:
@Betree : I've corrected the remaining issues and added some tests as well. Let me know how it looks and if anything else needs to be done. 😄

SudharakaP commented on server/lib/notifications.js:
```diff
@@ -215,9 +215,61 @@ const notifyUpdateSubscribers = async activity => {
+  if (service === 'youtube') {
+    return `https://img.youtube.com/vi/${id}/0.jpg`;
+  } else if (service === 'anchorFm') {
+    return `https://theme.zdassets.com/theme_assets/2009830/ed34a3258bf8d79c3db30e4269dd95052481fc50.png`;
```
@Betree : Good point. This image I put comes actually from the anchor.fm help page; https://help.anchor.fm/hc/en-us/articles/360016501412-Creating-your-podcast-cover-art. 

![image](https://user-images.githubusercontent.com/12435965/120493990-479a9a80-c370-11eb-99fc-7cdeabf6ad28.png)

Betree replied:
Yes

SudharakaP replied:
Done. 👍🏽

Betree commented on server/lib/notifications.js:
```diff
@@ -215,9 +215,61 @@ const notifyUpdateSubscribers = async activity => {
 };
 
+function replaceVideosByImagePreviews(activity) {
+  const iframePreviewRegex = /<iframe\s[^>]*src="([^"]+)"[^>]*><\/iframe>/gi;
```
I'm not comfortable with us using regexps. This one won't cover all use cases (what if the regex has no closing tags? or something between the tags (i.e. `<iframe>Hello</iframe>`)? I'm also uncertain about the `.replace`, it's difficult to rely on this.

I'd like us to use a real HTML parser to achieve this. Thankfully, `sanitize-html` has just what we need: https://github.com/apostrophecms/sanitize-html#transformations. By using the `transformTags` option we'll be able to replace iframes by images safely and the code should be easier to understand.

SudharakaP replied:
Thanks. I've used the transformations from `sanitize-html`. Let me know how it looks. I've enabled all content types we support except video embeds; 

https://github.com/opencollective/opencollective-api/blob/b7b4aef1c118a594255e1bab1ba4c4702fe117f4/server/lib/notifications.js#L245-L275

since we area dealing with a rich text editor text. 🤔

Betree commented on server/lib/notifications.js:
```diff
@@ -215,9 +215,61 @@ const notifyUpdateSubscribers = async activity => {
+
+function constructPreviewImageURL(service, id) {
+  if (service === 'youtube') {
+    return `https://img.youtube.com/vi/${id}/0.jpg`;
```
In the spirit of defensive programming, I think we should sanitize the `id` here. A YouTube id is `[a-zA-Z0-9_-]{11}`, if we don't get something like that we should return `null`.

SudharakaP replied:
Thanks. I've changed that to sanitize using the regex you suggested. 👍🏽

SudharakaP commented:
@Betree : Also in the final commit; 5d66ec0 I've removed one test which I've added. This isn't a YouTube link or a AnchorFM link and hence this kinda of iframe isn't valid and the API will not add such content at all. 

https://github.com/opencollective/opencollective-api/blob/f1648e68812a7d724e9bb7b6a29f744cd73debbc/server/lib/sanitize-html.ts#L90

Hence the test will always fail for such a test I think. 😉

Betree commented:
> @Betree : Also in the final commit; [5d66ec0](https://github.com/opencollective/opencollective-api/commit/5d66ec099401a097fe9b2f26a001650af02afe88) I've removed one test which I've added. This isn't a YouTube link or a AnchorFM link and hence this kinda of iframe isn't valid and the API will not add such content at all.
> 
> https://github.com/opencollective/opencollective-api/blob/f1648e68812a7d724e9bb7b6a29f744cd73debbc/server/lib/sanitize-html.ts#L90
> 
> Hence the test will always fail for such a test I think. wink

This test was there on purpose, the goal was to make sure iframe from other domains are stripped out. We should add the test back.

Betree commented on server/lib/notifications.js:
```diff
@@ -236,9 +238,77 @@ const notifyUpdateSubscribers = async activity => {
 };
 
+function replaceVideosByImagePreviews(activity) {
+  const sanitizeOptions = buildSanitizerOptions({
```
Would be great to import this from https://github.com/opencollective/opencollective-api/blob/0bc1022162162c6b2621eb6c6df8e897145ae93f/server/models/Update.js#L15, to make sure changes to options there are added here as well.

SudharakaP replied:
Corrected this one. 👍🏽

Betree reviewed (commented):
Looking great! Small feedback below. Let's then add back the test and we'll be good to go.

SudharakaP commented:
> Looking great! Small feedback below. Let's then add back the test and we'll be good to go.

Thanks for the explanation, that makes sense. I've added the test back and adjusted the count here,

https://github.com/opencollective/opencollective-api/blob/5b31336033ee042a36a03c3b6e439f1aa37aee95/test/server/lib/notifications.test.ts#L170-L174

so that we don't consider the final test (with `www.test.com`) as that one is stripped out. Let me know if everything looks good now to merge. 👍🏽

Betree commented on server/lib/notifications.js:
```diff
@@ -236,9 +238,68 @@ const notifyUpdateSubscribers = async activity => {
 };
 
+function replaceVideosByImagePreviews(activity) {
+  sanitizerOptions.transformTags = {
```
Doing so is mutating the `sanitizerOptions` from `../models/Update`.

I suggest to do something like:

```es6
import { sanitizerOptions as updateSanitizerOptions } from '../models/Update';

function replaceVideosByImagePreviews(activity) {
  const sanitizerOptions = {
    ...updateSanitizerOptions,
    ...
  }
```

SudharakaP replied:
Done. Thanks for pointing it out. 👍🏽

Betree reviewed (changes requested):
> so that we don't consider the final test (with www.test.com) as that one is stripped out. Let me know if everything looks good now to merge.

The `iframe` should be stripped out from the content, but the mail should still be sent

Betree reviewed (approved):
:rocket: :rocket: :rocket:
